src/vs/workbench/contrib/notebook/browser/contrib/outline/notebookOutline.ts

```diff
@@ -118,13 +142,109 @@ class NotebookOutlineRenderer implements ITreeRenderer<OutlineEntry, FuzzyScore,
 				template.container.style.setProperty('--outline-element-color', color?.toString() ?? 'inherit');
 			}
 		}
+
+		if (this._target === OutlineTarget.OutlinePane) {
+			const nbCell = node.element.cell;
+			const nbViewModel = this._editor?.getViewModel();
+			if (!nbViewModel) {
+				return;
+			}
+			const idx = nbViewModel.getCellIndex(nbCell);
+			const length = isCodeCell ? 0 : nbViewModel.getFoldedLength(idx);
+			const foldingState = isCodeCell ? CellFoldingState.None : ((nbCell as MarkupCellViewModel).foldingState);
+
+			const scopedContextKeyService = template.elementDisposables.add(this._contextKeyService.createScoped(template.container));
+			NotebookOutlineContext.CellKind.bindTo(scopedContextKeyService).set(isCodeCell ? CellKind.Code : CellKind.Markup);
+			NotebookOutlineContext.CellHasChildren.bindTo(scopedContextKeyService).set(length > 0);
+			NotebookOutlineContext.CellHasHeader.bindTo(scopedContextKeyService).set(node.element.level !== 7);
+			NotebookOutlineContext.CellFoldingState.bindTo(scopedContextKeyService).set(foldingState);
+			NotebookOutlineContext.OutlineElementTarget.bindTo(scopedContextKeyService).set(this._target);
+			this.setupFoldingListeners(nbViewModel, scopedContextKeyService, template, nbCell);
```

Comment:
It seems in `setupFoldingListeners` we force converting nbCell to MarkdownCellViewModel, however it can be CodeCell. My suggestion would be moving all folding state computation into the same function, i.e.,

```ts
setupFolding(cell: ICellViewModel, nbViewModel: INotebookViewModel, scopedContextKeyService: IContextKeyService, template: NotebookOutlineTemplate) {
	const foldingState = cell.cellKind === CellKind.Code ? CellFoldingState.None : ((nbCell as MarkupCellViewModel).foldingState);
	const foldingStateContextKey = NotebookOutlineContext.CellFoldingState.bindTo(scopedContextKeyService).set(foldingState);

	if (cell.cellKind === CellKind.Code) {
		template.elementDisposables.add(nbViewModel.onDidFoldingStateChanged(() => {
			const foldingState = (nbCell as MarkupCellViewModel).foldingState;
			foldingStateContextKey.set(foldingState);
		}));
	}
}

```